### PR TITLE
refactor(EventBuilder): Moved decision service and bucketing from EventBuilder to Optimizely 

### PR DESCRIPTION
**Pull request by AbdurRafay**
### Summary
- Moved decision service and bucketing from EventBuilder to Optimizely.
- Pulled eventTag stuff into another method.
- Revised activate & track methods to validate inputs.
- Revised test cases of Optimizely & EventBuilder.

### CODE REVIEW

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventParameterKeys.h [hunk 1]
```diff
@@ -48,7 +48,6 @@ extern NSString * const OPTLYEventParameterKeysFeaturesName;
 extern NSString * const OPTLYEventParameterKeysDecisionCampaignId;
 extern NSString * const OPTLYEventParameterKeysDecisionExperimentId;
 extern NSString * const OPTLYEventParameterKeysDecisionVariationId;
```
Comment (msohailhussain) [inline OptimizelySDKCore/OptimizelySDKCore/OPTLYEventParameterKeys.h:50]: update year in the header.

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventParameterKeys.m [hunk 2]
```diff
@@ -46,7 +46,6 @@
 NSString * const OPTLYEventParameterKeysDecisionCampaignId              = @"campaign_id";       // nonnull
 NSString * const OPTLYEventParameterKeysDecisionExperimentId            = @"experiment_id";     // nonnull
 NSString * const OPTLYEventParameterKeysDecisionVariationId             = @"variation_id";      // nonnull
-NSString * const OPTLYEventParameterKeysDecisionIsLayerHoldback         = @"is_campaign_holdback";
```
Comment (msohailhussain) [inline OptimizelySDKCore/OptimizelySDKCore/OPTLYEventParameterKeys.m:49]: update license header.

**Comment by msohailhussain** [comment]
There are too many changes in this PR, still reviewing it.

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.h [hunk 3]
```diff
@@ -34,39 +34,45 @@ NS_ASSUME_NONNULL_END
...
  * @param attributes A map of attribute names to current user attribute values.
  * @return A map of parameters for an impression event. This value can be nil.
  *
  */
-- (nullable NSDictionary *)buildImpressionEventTicket:(nonnull OPTLYProjectConfig *)config
-                                         userId:(nonnull NSString *)userId
-                                  experimentKey:(nonnull NSString *)experimentKey
-                                    variationId:(nonnull NSString *)variationId
-                                     attributes:(nullable NSDictionary<NSString *, NSString *> *)attributes;
+- (nullable NSDictionary *)buildImpressionEventTicketForUser:(nonnull NSString *)userId
```
Comment (msohailhussain) [inline OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.h]: Why did you rename it? 
What's the purpose to use Ticket in the name?

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.h [hunk 4]
```diff
@@ -34,39 +34,45 @@ NS_ASSUME_NONNULL_END
...
  * @return A map of parameters for a conversion event. This value can be nil.
  *
  */
-- (nullable NSDictionary *)buildConversionTicket:(nonnull OPTLYProjectConfig *)config
-                               bucketer:(nonnull id<OPTLYBucketer>)bucketer
-                                 userId:(nonnull NSString *)userId
-                              eventName:(nonnull NSString *)eventName
-                              eventTags:(nullable NSDictionary *)eventTags
-                             attributes:(nullable NSDictionary<NSString *, NSString *> *)attributes;
+- (nullable NSDictionary *)buildConversionEventTicketForUser:(nonnull NSString *)userId
```
Comment (msohailhussain) [inline OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.h]: Same comments.

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.m [hunk 5]
```diff
@@ -14,114 +14,113 @@
...
+                                        experiment:(OPTLYExperiment *)experiment
+                                         variation:(OPTLYVariation *)variation
+                                        attributes:(NSDictionary<NSString *,NSString *> *)attributes {
+    if (!self.config) {
         return nil;
     }
     
-    if ([experimentKey length] == 0) {
-        [config.logger logMessage:OPTLYLoggerMessagesExperimentKeyInvalid withLevel:OptimizelyLogLevelWarning];
+    if ([OPTLYEventBuilderDefault isEmptyString:userId]) {
```
Comment (msohailhussain) [inline OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.m]: I think these conditions must be placed before calling EventBuilder methods. it's an internal class that should be used only to generate payloads, not to check or validate conditions except attributes.

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.m [hunk 6]
```diff
@@ -14,114 +14,113 @@
...
+                                                              decisions:decisions eventTags:eventTags
+                                                             attributes:attributes];
     NSDictionary *conversionParams = [self createImpressionOrConversionParamsWithCommonParams:commonParams conversionOrImpressionOnlyParams:conversionOnlyParams];
     
     return conversionParams;
 }
 
 
-- (NSNumber *)revenueValue:(OPTLYProjectConfig *)config value:(NSObject *)value {
+- (NSNumber *)revenueValueFromEventTags:(NSDictionary *)eventTags {
```
Comment (msohailhussain) [inline OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.m]: can we use external alias in argument instead of FromTags in the method name.

Comment (msohailhussain) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.m]: @thomaszurkan-optimizely I think these two methods revenueValueFromEventTags and numericValue should be moved in Utils class and this builder class should call those methods. 
Shouldnt' be a part of this class. 
What do you suggest?

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.m [hunk 7]
```diff
@@ -220,185 +223,144 @@ - (NSNumber *)numericValue:(OPTLYProjectConfig *)config value:(NSObject *)value
...
-        [config.logger logMessage:OPTLYLoggerMessagesNumericValueInvalid withLevel:OptimizelyLogLevelWarning];
+        [self.config.logger logMessage:OPTLYLoggerMessagesNumericValueInvalid withLevel:OptimizelyLogLevelWarning];
     };
     return answer;
 }
 
-- (NSDictionary *)createCommonParams:(OPTLYProjectConfig *)config userId:(NSString *)userId
-                          attributes:(NSDictionary<NSString *, NSString *> *)attributes {
-    NSMutableDictionary *params = [NSMutableDictionary new];
+- (NSDictionary *)createCommonParamsForUser:(NSString *)userId
```
Comment (msohailhussain) [inline OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.m]: not reviewing for now, will check it via fullstack-suite. Assuming you copied from the PR to correct EventBuilder payload.

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYLoggerMessages.h [hunk 8]
```diff
@@ -26,7 +26,16 @@ extern NSString *const OPTLYLoggerMessagesVariationUserAssigned;
 // info
 extern NSString *const OPTLYLoggerMessagesActivationSuccess;
 extern NSString *const OPTLYLoggerMessagesConversionSuccess;
+extern NSString *const OPTLYLoggerMessagesConversionFailure;
 // error
+extern NSString *const OPTLYLoggerMessagesUserIdInvalid;
```
Comment (msohailhussain) [inline OptimizelySDKCore/OptimizelySDKCore/OPTLYLoggerMessages.h:31]: We just refactored the code, why these constants are added, any log or case were missing in the SDK.?

File: OptimizelySDKCore/OptimizelySDKCore/Optimizely.m [hunk 9]
```diff
@@ -880,18 +866,69 @@ - (OPTLYVariation *)sendImpressionEventFor:(OPTLYExperiment *)experiment
...
 
+/**
+ Helper method to retrieve decisions for the user for experiment(s) using the provided event.
+
+ @param event The event which needs to be recorded.
+ @param userId Id for user.
+ @param attributes The user's attributes.
+ @return Array of dictionaries containing valid experiment Ids, variation Ids and layer Ids into which the user is bucketed.
+ */
+- (NSArray<NSDictionary *> *)decisionsFor:(OPTLYEvent *)event
```
Comment (msohailhussain) [inline OptimizelySDKCore/OptimizelySDKCore/Optimizely.m:883]: I didn't get the context of this method.

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.h (hunk 3 above)
Comment (AbdurRafay) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.h]: Since I made ProjectConfig as EventBuilder property to avoid passing in every method, I had to rename it. Ticket is already used in the method name.

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.h (hunk 4 above)
Comment (AbdurRafay) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.h]: replied above.

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.h [hunk 10]
```diff
@@ -34,39 +34,45 @@ NS_ASSUME_NONNULL_END
...
+                                                       event:(nonnull OPTLYEvent *)event
+                                                   decisions:(nonnull NSArray<NSDictionary *> *)decisions
+                                                   eventTags:(nullable NSDictionary *)eventTags
+                                                  attributes:(nullable NSDictionary<NSString *, NSString *> *)attributes;
 @end
 
 @interface OPTLYEventBuilderDefault : NSObject<OPTLYEventBuilder>
 
+/// init is disabled. Please use initWithConfig to create an Event Builder
```
[the remark this answers is not in the document]
Comment (AbdurRafay) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.h:68]: EventBuilder needs to initialize with ProjectConfig so forcing to use initWithConfig.

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.m (hunk 7 above)
Comment (AbdurRafay) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.m]: Just removed ProjectConfig from the method signature.

File: OptimizelySDKCore/OptimizelySDKCore/Optimizely.m (hunk 9 above)
Comment (AbdurRafay) [reply OptimizelySDKCore/OptimizelySDKCore/Optimizely.m:883]: Helper method to retrieve decisions for the user from the provided event.

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.h (hunk 3 above)
Comment (mikeproeng37) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.h]: Since we are renaming, let's remove `ticket` from the name as none of the other SDKs call it that :)

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.m (hunk 6 above)
Comment (mikeproeng37) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.m]: Agreed. The other SDKs should already be doing that

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.m (hunk 7 above)
Comment (mikeproeng37) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.m]: Please also do some manual E2E on this PR and ensure that events are getting converted correctly and show up on the results page. That gives us much better guarantee over the compat-suite. Only asking for this because refactoring event builder is always risky :)

File: OptimizelySDKCore/OptimizelySDKCore/Optimizely.m [hunk 11]
```diff
@@ -112,52 +112,61 @@ - (OPTLYVariation *)activate:(NSString *)experimentKey
                   attributes:(NSDictionary<NSString *,NSString *> *)attributes
                     callback:(void (^)(NSError *))callback {
     
-    // get variation
-    OPTLYVariation *variation = [self variation:experimentKey userId:userId attributes:attributes];
+    __weak void (^_callback)(NSError *) = callback ? : ^(NSError *error) {};
+    
+    if ([Optimizely isEmptyString:experimentKey]) {
```
Comment (mikeproeng37) [inline OptimizelySDKCore/OptimizelySDKCore/Optimizely.m:117]: I'd rather we make this into another PR to keep this one from being longer than it already is

File: OptimizelySDKCore/OptimizelySDKCore/Optimizely.m [hunk 12]
```diff
@@ -394,22 +403,40 @@ - (void)track:(NSString *)eventKey
...
         return;
     }
     
-    NSDictionary *conversionEventParams = [self.eventBuilder buildConversionTicket:self.config
-                                                                          bucketer:self.bucketer
-                                                                            userId:userId
-                                                                         eventName:eventKey
-                                                                         eventTags:eventTags
-                                                                        attributes:attributes];
+    NSArray *decisions = [self decisionsFor:event userId:userId attributes:attributes];
```
Comment (mikeproeng37) [inline OptimizelySDKCore/OptimizelySDKCore/Optimizely.m:424]: Not sure what this is about, can you explain a bit more? It's a bit confusing that this happens in the top-level Optimizely class.

**Review by mikeproeng37 — CHANGES_REQUESTED** [review]

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.m (hunk 7 above)
Comment (msohailhussain) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYEventBuilder.m]: Sure Mike, I will take care of it.

File: OptimizelySDKCore/OptimizelySDKCore/Optimizely.m (hunk 12 above)
Comment (AbdurRafay) [reply OptimizelySDKCore/OptimizelySDKCore/Optimizely.m:424]: This is about making decisions array containing experiments and their bucketed variations which are associated with an event.
Purpose is to have the decision service and bucketing called within the optimizely object and not the event builder.

File: OptimizelySDKCore/OptimizelySDKCore/Optimizely.m (hunk 11 above)
Comment (AbdurRafay) [reply OptimizelySDKCore/OptimizelySDKCore/Optimizely.m:117]: These input validations were missing from Activate and Track APIs

**Comment by msohailhussain** [comment]
build

**Review by mikeproeng37 — COMMENTED** [review]
lgtm

**Review by mikeproeng37 — APPROVED** [review]
